lua/telescope/actions/history.lua

```diff
@@ -0,0 +1,218 @@
+        -- to 50. Now we need to remove the no longer needed items.
+        -- We could do this check on startup as well but currently i don't think its
+        -- necessary.
+        local diff = len - self.limit
+        for i = diff, 1, -1 do
+          table.remove(self.content, i)
+        end
+        write_async(self.path, table.concat(self.content, '\n') .. '\n', 'w')
+      else
+        append_async(self.path, line .. '\n')
+      end
+    end
+  end
+  self:reset()
+end
+
+function SimpleHistory:get_next()
+  if not self.enabled then return end
+
+  local next_idx = self.index + 1
+  if next_idx <= #self.content then
+    self.index = next_idx
+    return self.content[next_idx]
+  end
+  self.index = #self.content + 1
+  return nil
+end
+
+function SimpleHistory:get_prev()
+  if not self.enabled then return end
+
+  local next_idx = self.index - 1
+  if next_idx >= 1 then
+    self.index = next_idx
+    return self.content[next_idx]
+  end
+  return nil
+end
+
+m.get_simple_history = function(...)
+  return SimpleHistory:new(...)
+end
+
+------------------------------------------------------------------------------
+------------------------------------------------------------------------------
+-- DB HISTORY SCRATCH. MOVING TO EXTENSION AFTER MAKING IT WORK --------------
+
+local sql = require('sql')
+
+-- TODO(conni2461): Serialize metadata
+-- TODO(conni2461): CURRENTLY HARDCODED picker and cwd
+local SmartHistory = {}
+SmartHistory.__index = SmartHistory
+
+function SmartHistory:new()
+  local obj = {
+    enabled = true
+  }
+  if not conf.history_location then
+    obj.enabled = false
+    return setmetatable(obj, self)
+  end
+  if conf.history_limit then
+    obj.limit = conf.history_limit
+  end
+
+  obj.path = vim.fn.expand(conf.history_location)
+  obj.db = sql:open(obj.path)
+  -- any chance i can get a table object here?!?
+  -- because using obj.db:table to create one feels weird because i can't pass in
+  -- a schema at the same time. WTF tami am i just stupid.
+  -- I think the readme is confusing LULW
+  obj.db:create("histories", {
+    id = { "integer", "primary", "key" },
+    content = "text",
+    picker = "text",
+    cwd = "text",
+    ensure = true
+  })
+
+  -- we are using this all the time
+  -- FIRST QUERY THEN WORKING WITH THIS HERE
+  obj.current = nil
+  obj.index = nil
+
+  return setmetatable(obj, self)
+end
+
+function SmartHistory:reset()
+  if not self.enabled then return end
+  self.current = nil
+  self.index = nil
+  collectgarbage()
+end
+
+function SmartHistory:append(line)
+  if not self.enabled then return end
+
+  if line ~= '' then
+    self:__ensure_current('find_files', vim.loop.cwd())
+    if self.current[#self.current].content ~= line then
+      -- TODO(conni2461): HANDLE LIMIT. REMOVE first n with id
+      self.db:insert("histories", {
+        content = line,
+        picker = "find_files",
+        cwd = vim.loop.cwd()
+      })
```

Comment:
with using table module this becomes:

```lua
self.data:insert { ... } 
```

btw if you need to append json, just define the field type to be json and sql.nvim will serialize and deserialize lua table automatically. might fail though on functions not sure

Reply:
That is exactly waht i wanted to do but when using `db:table("history", { schema = { ... } })` it wasn't added :thinking: Maybe i had a typo somewhere. i try it again